### PR TITLE
Terraform

### PR DESCRIPTION
**Pull request by alabbahs**
This is the terraform branch including all the terraform files for the individual AWS resources being used to in our architecture. The list of the resources are as follows...

- Fargate task to host the streamlit dashboard, running a dockerised image version of the Streamlit app.
-  Lambda solution deploying our live pipeline, which runs every two hours between 09:00-17:00 and uploads newly added data from the archives website into our database.
- S3 bucket which temporarily stores GPT prompt responses during live pipeline insertion of data.
-  Postgresql database which stores the primary data being cleaned, processed and output.

### CODE REVIEW

File: dashboard/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+COPY requirements.txt .
+
+RUN pip install --no-cache-dir -r requirements.txt
+
+COPY . .
+
+EXPOSE 8501
+
+CMD ["streamlit", "run", "dashboard.py", "--server.port=8501", "--server.address=0.0.0.0"]
```
Comment (LinfanS) [inline dashboard/Dockerfile:13]: Does server address need to be specified here? Does it not mean localhost, does this still work when deployed with a public IP?

File: terraform/fargate.tf [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        Principal = {
+          Service = "ecs-tasks.amazonaws.com"
+        },
+        Action = "sts:AssumeRole"
+      }
+    ]
+  })
+
+  managed_policy_arns = [
+    "arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy"
```
Comment (LinfanS) [inline terraform/fargate.tf:64]: Nice of you to use the predefined policy.

File: terraform/lambda.tf [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+resource "aws_scheduler_schedule" "court_lambda_schedule" {
+  name       = "c12-court-lambda-schedule"
+  group_name = "default"
+
+  flexible_time_window {
+    mode = "OFF"
+  }
+
+  schedule_expression = "cron(0 9-17/2 ? * * *)"
```
Comment (LinfanS) [inline terraform/lambda.tf:118]: Thanks for updating the cron to run everyday as discussed.

File: terraform/lambda.tf [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    ]
+  })
+}
+
+resource "aws_cloudwatch_log_group" "lambda_log_group" {
+  name              = "/aws/lambda/c12-court-pipeline"
+  retention_in_days = 14
+}
+
+resource "aws_iam_role_policy" "lambda_logging_policy" {
```
Comment (LinfanS) [inline terraform/lambda.tf:25]: Thanks for updating the lambda with cloudwatch logging.

File: terraform/main.tf [hunk 5]
```diff
@@ -0,0 +1,5 @@
+provider "aws" {
```
Comment (LinfanS) [inline terraform/main.tf:1]: Are the contents of this file actually doing anything at the moment? Is the main.tf file necessary?

File: terraform/lambda.tf [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        }
+      ]
+    })
+  }
+
+  depends_on = [aws_lambda_function.my_lambda_function]
+}
+
+resource "aws_scheduler_schedule" "court_lambda_schedule" {
+  name       = "c12-court-lambda-schedule"
```
Comment (LinfanS) [inline terraform/lambda.tf:111]: Is an argument required here to give the lambda function an input of empty json e.g. {}

File: terraform/s3.tf [hunk 7]
```diff
@@ -0,0 +1,10 @@
+resource "aws_s3_bucket" "c12-court-s3" {
+  bucket = "c12-court-bucket"
+
+  force_destroy = true
+
+  tags = {
```
Comment (LinfanS) [inline terraform/s3.tf:6]: Are the tags necessary here? What do they do?

**Review by LinfanS — APPROVED** [review]
Additionally,  it might make more sense to separate the .tf files by what they are creating such as dashboard/live_pipeline/api instead of by the type of cloud infrastructure, makes it easier to locate and debug individual parts of the architecture when a project has many different parts. Approving this pull request as we've already tested it works. Nice work Hamoodi.

**Review by annac02 — APPROVED** [review]

File: dashboard/Dockerfile (hunk 1 above)
Comment (alabbahs) [reply dashboard/Dockerfile]: Yes, we discussed this and came to an agreement as this server address allows you to access it from any address, localhost or other. It does work as it's currently deployed

File: terraform/main.tf (hunk 5 above)
Comment (alabbahs) [reply terraform/main.tf:1]: The provider has to be defined somewhere, if not here then in on of the random resources I picked and I didn't want that place to be arbitrary such as, defining the provider in the db.tf. So main.tf made sense.

File: terraform/lambda.tf (hunk 6 above)
Comment (alabbahs) [reply terraform/lambda.tf:120]: It's not necessary when setting it up in terraform, perhaps when testing it and simulating an invocation.

File: terraform/s3.tf (hunk 7 above)
Comment (alabbahs) [reply terraform/s3.tf:6]: Not necessary but, I thought it was a good practice given identifying the s3 was crucial.

**Pull request merged**